include/onnxruntime/core/framework/allocator.h

```diff
@@ -12,10 +12,27 @@
 // This configures the arena based allocator used by ORT
 // See docs/C_API.md for details on what these mean and how to choose these values
 struct OrtArenaCfg {
-  size_t max_mem;                // use 0 to allow ORT to choose the default
-  int arena_extend_strategy;     // use -1 to allow ORT to choose the default, 0 = kNextPowerOfTwo, 1 = kSameAsRequested
-  int initial_chunk_size_bytes;  // use -1 to allow ORT to choose the default
-  int max_dead_bytes_per_chunk;  // use -1 to allow ORT to choose the default
+  OrtArenaCfg() : max_mem(0),
+                  arena_extend_strategy(-1),
+                  initial_chunk_size_bytes(-1),
+                  max_dead_bytes_per_chunk(-1),
+                  initial_regrowth_chunk_size_bytes_after_shrink(-1),
+                  shrink_on_every_run(false) {}
+  OrtArenaCfg(size_t max_mem, int arena_extend_strategy, int initial_chunk_size_bytes,
+              int max_dead_bytes_per_chunk, int initial_regrowth_chunk_size_bytes_after_shrink,
+              bool shrink_on_every_run) : max_mem(max_mem),
+                                          arena_extend_strategy(arena_extend_strategy),
+                                          initial_chunk_size_bytes(initial_chunk_size_bytes),
+                                          max_dead_bytes_per_chunk(max_dead_bytes_per_chunk),
+                                          initial_regrowth_chunk_size_bytes_after_shrink(initial_regrowth_chunk_size_bytes_after_shrink),
+                                          shrink_on_every_run(shrink_on_every_run) {}
+
+  size_t max_mem;                                      // use 0 to allow ORT to choose the default
+  int arena_extend_strategy;                           // use -1 to allow ORT to choose the default, 0 = kNextPowerOfTwo, 1 = kSameAsRequested
+  int initial_chunk_size_bytes;                        // use -1 to allow ORT to choose the default
+  int max_dead_bytes_per_chunk;                        // use -1 to allow ORT to choose the default
+  int initial_regrowth_chunk_size_bytes_after_shrink;  // use -1 to allow ORT to choose the default
+  bool shrink_on_every_run;                            // default `false`
 };
```

Comment:
Should shrink_on_every_run be an option at the inference session level (either coming from SessionOptions or RunOptions) and the arena just provides the Shrink mechanism?

Reply:
+1 for making it a RunOption since that's when Shrink() is invoked.

Reply:
What if we wanted to only shrink some memory arenas ? This would have to be configured in the arena configuration of individual arenas itself (as opposed to have a "global" flag in session options/ inference session which removes the option of shrinking only a subset of all possible arenas involved in that specific session).

Reply:
By "some memory arenas" - I mean let us say users would like to control the growth of just the cuda memory arena but are quite happy to let the cpu memory arena grow (without shrinkage) as they may feel its growth rate is quite minimal and they can avoid the cost of shrinkage and re-allocation down the line. 

To support this, we will need to have a flag in RunOptions per EP to provide this granular control. I deemed it may be cleaner to have this as part of the arena configuration itself.

Reply:
I think two settings may make more sense. An arena could enable/disable shrinking (a call to Shrink is a no-op if disabled) to provide per-arena control. And a run or session option could enabled/disable the 'shrink post-Run' functionality (which determines whether any calls to Shrink are made).

Reply:
Initially (a few days back) I'd suggested keeping the shrink config outside the arena. It's the callers responsibility to call or skip calling Shrink which can be controlled based on just one config - a run level option. Less knobs, less confusion.

Reply:
Moving the flag to shrink the arena to RunOptions is a good suggestion and as it really is a run option. The only question is how do make this flag granular in that we support turning this ON/OFF for individual allocators ?

We can have a flag in the arena config that enables/disables shrinking for the individual arena and we can have one flag in RunOptions for the user to decide to choose to shrink but this opens up possibilities for confusion as well. The user may not have configured the arena to shrink but might set the RunOption flag to trigger the shrinking (which will result in a no-op) and the user might wonder about this. 

I understand that there is some reservation about the word "Run" being used in the context of an arena config (as an arena shouldn't know about Run) so I can move this boolean flag to the EP option (ORtCudaProviderOptions). A flag here can be set by the user to trigger an arena shrink at the end of every Run (The EP knows about Runs and it looks less out of place here) and the arena config can have the boolean flag removed. Thoughts ?

Reply:
I missed your earlier comment. If we want to enable selective shrinkage, I still feel it's better to have them configured as part of RunOptions. Having the shrink_on_run in the arena config is a bit misleading in that it gives the indication that the arena will get automatically shrunk on some Run call which happens outside the context. Moreover, I think having to enable 2 configs to cause the shrinkage is a source of confusion (if I've enabled shrinkage once, why should I enable it again and which one should take precedence). I suggest we go with device level RunOptions config like: 'enable_arena_shrinkage': 'cpu', 'gpu:0', 'gpu:1'. We're any way storing the allocators in session state by device id.

Reply:
I like the idea of per-allocator values in the RunOptions. Only concern is the naming. 

Technically we store the allocators in session state, but they can be retrieved using OrtMemoryInfo (which actually ignores the OrtDevice field) where the allocator name is a key factor (and tends to be set to the EP name for minimal re-use), or they can be retrieved using OrtDevice where the device type is the key factor. We really need to rationalize the whole allocator setup...

We use the former from the ExecutionFrame where a value has an OrtMemoryInfo for the EP running the node, and the latter when copying graph inputs/outputs across devices.

Ideally this whole setup would be device centric, and something like 'gpu:0' in the config would be simple and make sense, although there's an obvious edge case of one CUDA GPU and one ARM GPU on the same machine (whether you need to support different shrink settings for both is a different question).

Maybe we need some custom logic short term that makes this setting device centric by iterating all the allocators in session state (lookup by device will just return the first matching one) and matches the OrtDevice value for the allocator.

Reply:
rather than go gpu:0, gpu:1 - how about going with something like cuda:0 in the short term then ?

Reply:
I was thinking 'gpu' was fine but it needs to match against OrtDevice.DeviceType and probably be limited to a memory type of default for now (unless you specify the OrtDevice.MemoryType as well). 

'cuda' is murkier as I think that would have to match the string (or prefix) in the allocator name in order to be applied (as well as info in OrtDevice).

Reply:
Made the shrink enabler part of RunOptions and fixed all the relevant plumbing